### PR TITLE
support for parallel H5MD dump and restore

### LINKED ISSUE #346: Check if hdf5 is parallel at CMake time
Currently everything will build, but then the test fails with an error that hdf5 is not parallel

### PR DESCRIPTION
**Pull request by XzzX**
closes #346
fix #356

### CODE REVIEW

File: testsuite/io/test_h5md_parallel.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python3
+#  Copyright (C) 2015-2017(H)
```
Comment (junghans) [inline testsuite/io/test_h5md_parallel.py:2]: Huh?

File: src/io/hdf5.hpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/*
+  This file is part of ESPResSo++.
```
Comment (junghans) [inline src/io/hdf5.hpp:2]: years missing

File: testsuite/io/test_h5md_parallel.py (hunk 1 above)
Comment (XzzX) [reply testsuite/io/test_h5md_parallel.py:2]: Copied from one of the existing test cases.

File: src/io/hdf5.hpp (hunk 2 above)
Comment (XzzX) [reply src/io/hdf5.hpp:2]: The header is very inconsistent throughout `ESPResSo++`. How do you want it to be?

File: src/io/DumpH5MDParallel.cpp [hunk 3]
```diff
@@ -0,0 +1,322 @@
+/*
+  This file is part of ESPResSo++.
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:2]: Let's stick to this kind of header
```
  Copyright (C) 2021
      Sebastian Eibl

  This file is part of ESPResSo++.
```

File: src/io/DumpH5MDParallel.cpp [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    auto dstSpace = CHECK_HDF5(H5Dget_space(dataset));
+    CHECK_HDF5(H5Sselect_hyperslab(dstSpace, H5S_SELECT_SET, offset.data(), stride.data(),
+                                   count.data(), localDims.data()));
+
+    std::vector<hsize_t> localOffset(globalDims.size(), 0);
+    auto srcSpace = CHECK_HDF5(H5Screate_simple(localDims.size(), localDims.data(), NULL));
+    CHECK_HDF5(H5Sselect_hyperslab(srcSpace, H5S_SELECT_SET, localOffset.data(), stride.data(),
+                                   count.data(), localDims.data()));
+
+    auto datawrite_id = CHECK_HDF5(H5Pcreate(H5P_DATASET_XFER));
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:57]: Are there any particular reason to use snake case?

File: src/io/DumpH5MDParallel.cpp [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+
+    CHECK_HDF5(H5Pclose(datawrite_id));
+    CHECK_HDF5(H5Sclose(dstSpace));
+    CHECK_HDF5(H5Sclose(srcSpace));
+    CHECK_HDF5(H5Dclose(dataset));
+    CHECK_HDF5(H5Sclose(dataspace));
+}
+
+void DumpH5MDParallel::writeHeader(hid_t fileId) {
+    auto group1 = H5Gcreate(fileId, "/h5md", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:69]: Shouldn't this be also wrap with `CHECK_HDF5`?

File: src/io/DumpH5MDParallel.cpp [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+}
+
+void DumpH5MDParallel::writeBox(hid_t fileId) {
+    std::string groupName = "/particles/atoms/box";
+    auto group = CHECK_HDF5(H5Gcreate(fileId, groupName.c_str(), H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT));
+    std::vector<int> dims = {3};
+    CHECK_HDF5(H5LTset_attribute_int(fileId, groupName.c_str(), "dimension", dims.data(), dims.size()));
+
+    auto boundaryType = H5Tcopy(H5T_C_S1);
+    CHECK_HDF5( H5Tset_size(boundaryType, 8) );
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:95]: Unnecessary space
```suggestion
    CHECK_HDF5(H5Tset_size(boundaryType, 8));
```

File: src/io/DumpH5MDParallel.cpp [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+    CHECK_HDF5( H5Tset_strpad( boundaryType, H5T_STR_NULLPAD ));
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:96]:
```suggestion
    CHECK_HDF5(H5Tset_strpad( boundaryType, H5T_STR_NULLPAD));
```

File: src/io/DumpH5MDParallel.cpp [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+    std::vector<hsize_t> boundaryDims = {3};
+    auto space = H5Screate_simple(boundaryDims.size(), boundaryDims.data(), nullptr);
+    auto att = H5Acreate(group, "boundary", boundaryType, space, H5P_DEFAULT, H5P_DEFAULT);
+    CHECK_HDF5(H5Awrite(att, boundaryType, "periodicperiodicperiodic"));
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:100]: Why do you have to  put here `periodiperiodicperiodic` ?

File: src/io/DumpH5MDParallel.cpp [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+
+    std::vector<double> edges = {system_->bc->getBoxL()[0], system_->bc->getBoxL()[1], system_->bc->getBoxL()[2]};
+    CHECK_HDF5(H5LTset_attribute_double(fileId, groupName.c_str(), "edges", edges.data(), edges.size()));
+
+    CHECK_HDF5(H5Gclose(group));
+}
+
+
+void DumpH5MDParallel::writeId(hid_t fileId) {
+    std::string groupName = "/particles/atoms/"+idDataset;
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:113]:
```suggestion
    std::string groupName = "/particles/atoms/" + idDataset;
```

File: src/io/DumpH5MDParallel.cpp [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+    auto group =
+        H5Gcreate(fileId, groupName.c_str(), H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+
+    std::vector<int64_t> data;
+    data.reserve(numLocalParticles * 1);
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:118]: Do you need this multiplication by `1` ?

File: src/io/DumpH5MDParallel.cpp [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+    for (iterator::CellListIterator cit(system_->storage->getRealCells()); !cit.isDone(); ++cit) {
+        data.emplace_back(cit->id());
+    }
+    CHECK_EQUAL(data.size(), numLocalParticles * 1);
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:122]: As well here? or this is only for implicit type conversion?

File: src/io/RestoreH5MDParallel.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
+# Copyright (c) 2015-2018
```
Comment (jkrajniak) [inline src/io/RestoreH5MDParallel.py:1]: Please update the header, put your first, last name and institute (if applicable)

**Review by jkrajniak — DISMISSED** [review]

File: src/io/DumpH5MDParallel.cpp (hunk 3 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:2]: done

File: src/io/DumpH5MDParallel.cpp (hunk 4 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:57]: no, corrected.

File: src/io/DumpH5MDParallel.cpp (hunk 5 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:69]: indeed, fixed.

File: src/io/DumpH5MDParallel.cpp (hunk 6 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:95]: fixed

File: src/io/DumpH5MDParallel.cpp (hunk 7 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:96]: fixed

File: src/io/DumpH5MDParallel.cpp (hunk 9 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:113]: fixed.

File: src/io/RestoreH5MDParallel.py (hunk 12 above)
Comment (XzzX) [reply src/io/RestoreH5MDParallel.py:1]: fixed

File: src/io/DumpH5MDParallel.cpp (hunk 10 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:118]: I am generating all these functions and this is the number of dimensions. I will do this more clearly as I think this also improves readability of the code.

File: src/io/DumpH5MDParallel.cpp (hunk 8 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:100]: Judging from
https://github.com/espressopp/espressopp/blob/a967dada2b0fda6d390b89abcdc1f1541068d823/src/io/DumpH5MD.py#L216-L227
this seems to be the only supported case.

**Review by fedepad — COMMENTED** [review]
I made a small comment on the binary compare.

File: testsuite/io/test_h5md_parallel.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        dump_h5md_parallel.dump()
+
+        self.system.storage.removeAllParticles()
+
+        restore_h5md_parallel = espressopp.io.RestoreH5MDParallel(self.system, 'reference.h5')
+        restore_h5md_parallel.restore()
+        dump_h5md_parallel = espressopp.io.DumpH5MDParallel(self.system, 'dump.h5')
+        dump_h5md_parallel.dump()
+
+        self.assertTrue(binary_compare('reference.h5', 'dump.h5'))
```
[the remark this answers is not in the document]
Comment (XzzX) [reply testsuite/io/test_h5md_parallel.py:55]: I initially used `h5diff`, however, its a little bit tricky to parse the results correctly. Some errors are returned as return codes, some are printed to `stdout` and some to `stderr`. Therefore, I decided to go with the custom binary compare.

As this is a test case which runs often the file size is kept to a minimum. No parallel comparison is needed here.

File: testsuite/io/test_h5md_parallel.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+        ls_dmp = (subprocess.run(['h5ls', '-r', 'reference.h5'], capture_output=True)).stdout
```
Comment (junghans) [inline testsuite/io/test_h5md_parallel.py:57]: Oh man, is there no way to do this internally.

Otherwise CMake will need to check for `h5ls`.

Comment (XzzX) [reply testsuite/io/test_h5md_parallel.py:57]: Sure I can depend on `h5py` or `pyh5md`. I do not want to reimplement a `h5` reader.

File: CMakeLists.txt [hunk 15]
```diff
@@ -1,6 +1,6 @@
 cmake_minimum_required(VERSION 3.12)
 
-project(espressopp LANGUAGES CXX)
+project(espressopp LANGUAGES C CXX)
```
Comment (junghans) [inline CMakeLists.txt:3]: Instead of adding `C` here, I would add `enable_language(C)` (with a small comment) where `find_package(HFD5)` is called.

File: src/CMakeLists.txt [hunk 16]
```diff
@@ -56,6 +56,10 @@ find_package(Threads REQUIRED)
 set(Boost_USE_MULTITHREADED ON)
 
 add_library(_espressopp ${ESPRESSO_SOURCES})
+#target_link_libraries( _espressopp PRIVATE hdf5::hdf5 hdf5::hdf5_hl ) requires CMake 3.19 :(
```
Comment (junghans) [inline src/CMakeLists.txt:59]: We should consider bumping to 3.19 in the CMake refactor.

File: src/io/DumpH5MDParallel.cpp (hunk 8 above)
Comment (jkrajniak) [reply src/io/DumpH5MDParallel.cpp:103]: So this `periodicperiodicperiodic` will be written as three element array?

**Review by fedepad — COMMENTED** [review]
Since HDF5 is a required dependency by this PR and once merged as seen in the `CMakeLists.txt`, the `README.md` should be updated to include HDF5 as a dependency like the others dependencies that you cite (Boost, etc.).
Unless I am missing it somehow, I don't see the `README.md` in the list of changed files.

File: CMakeLists.txt (hunk 15 above)
Comment (XzzX) [reply CMakeLists.txt:3]: What is the benefit of this?

File: src/CMakeLists.txt (hunk 16 above)
Comment (XzzX) [reply src/CMakeLists.txt:59]: In principle I agree. However, currently, CMake 3.19 must be installed explicitly as it is quite new.

**Comment by XzzX** [comment]
> Since HDF5 is a required dependency by this PR and once merged as seen in the `CMakeLists.txt`, the `README.md` should be updated to include HDF5 as a dependency like the others dependencies that you cite (Boost, etc.).
> Unless I am missing it somehow, I don't see the `README.md` in the list of changed files.

`Readme.md` already contains `HDF5 library (>= 1.8) (optional for H5MD file format)`. Currently it is neither looked for nor used, but someone already included it...

File: src/io/DumpH5MDParallel.cpp (hunk 8 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:103]: yes

File: CMakeLists.txt (hunk 15 above)
Comment (junghans) [reply CMakeLists.txt:3]: I think that would make it more clear that the only reason we  need `C` is due to `HDF5`.

File: testsuite/io/test_h5md_parallel.py (hunk 14 above)
Comment (XzzX) [reply testsuite/io/test_h5md_parallel.py]: I now require `h5py`. I think this is the best way.

File: CMakeLists.txt (hunk 15 above)
Comment (XzzX) [reply CMakeLists.txt:3]: ..., `C` is only needed due to a compile test within `FindHDF5.cmake`. The file extension is `c` and, therefore, a c compiler is needed. If we change that to `cpp` we do not need `C` as a language in our project. However, we have to ship our own slightly modified version of `FindHDF5.cmake`. At the same time this would allow us to use exported targets, since we can build upon the CMake 3.20 version. @junghans What do you think?

Comment (junghans) [reply CMakeLists.txt:3]: I like that! And we should report that issue to cmake upstream as well!

File: src/io/DumpH5MDParallel.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+    def __init__(self, system, filename):
+        cxxinit(self, io_DumpH5MDParallel, system, filename)
+
+    def dump(self):
+        if not (pmi._PMIComm and pmi._PMIComm.isActive() ) or pmi._MPIcomm.rank in pmi._PMIComm.getMPIcpugroup():
+            self.cxxclass.dump(self)
+
+
+
+if pmi.isController:
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.py:34]: Too many lines before

File: src/io/DumpH5MDParallel.cpp [hunk 18]
```diff
@@ -0,0 +1,479 @@
...
+        .def_readwrite("author", &DumpH5MDParallel::author)
+        .def_readwrite("dumpId", &DumpH5MDParallel::dumpId)
+        .def_readwrite("dumpType", &DumpH5MDParallel::dumpType)
+        .def_readwrite("dumpMass", &DumpH5MDParallel::dumpMass)
+        .def_readwrite("dumpQ", &DumpH5MDParallel::dumpQ)
+        .def_readwrite("dumpGhost", &DumpH5MDParallel::dumpGhost)
+        .def_readwrite("dumpPosition", &DumpH5MDParallel::dumpPosition)
+        .def_readwrite("dumpVelocity", &DumpH5MDParallel::dumpVelocity)
+        .def_readwrite("dumpForce", &DumpH5MDParallel::dumpForce)
+        .def_readwrite("idDataset", &DumpH5MDParallel::idDataset)
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:468]: If this writer/restorer is for H5MD format then we shouldn't allow to modify the name of datasets as they are defined in the file format standard. See here: https://nongnu.org/h5md/h5md.html#particles-group
I would remove these properties or keep them as read-only.

File: doc/sphinx/espressopp.io.DumpH5MDParallel.rst [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+---------------
+
+Dumping a simulation with ``espressopp.io.DumpH5MDParallel``
+
+::
+
+    dump_h5md_parallel = espressopp.io.DumpH5MDParallel(system, 'dump.h5')
+    dump_h5md_parallel.dump()
+
+
```
Comment (jkrajniak) [inline doc/sphinx/espressopp.io.DumpH5MDParallel.rst:19]: Let's add here 
1. the possible options on what can be stored and how to enable/disable it.
2. the path to the particles location as it is fixed (`/particles/atoms/`)

I would also add *Attention* or *Note* with information that the dump is valid only for V=const (NVT, NVE) simulation as we do not save box dimensions at every simulation step.

File: src/io/DumpH5MDParallel.cpp [hunk 20]
```diff
@@ -0,0 +1,479 @@
...
+    auto group1 = CHECK_HDF5(H5Gcreate(fileId, "/h5md", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT));
+    auto group2 =
+        CHECK_HDF5(H5Gcreate(fileId, "/h5md/author", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT));
+    auto group3 =
+        CHECK_HDF5(H5Gcreate(fileId, "/h5md/creator", H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT));
+    CHECK_HDF5(H5Gclose(group1));
+    CHECK_HDF5(H5Gclose(group2));
+    CHECK_HDF5(H5Gclose(group3));
+
+    std::vector<int> data = {2, 0};
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:88]: AFAIK the last version of H5MD is 1.1 (right @pdebuyl ?)

File: src/io/DumpH5MDParallel.cpp [hunk 21]
```diff
@@ -0,0 +1,479 @@
...
+
+    CHECK_HDF5(H5Gclose(group));
+}
+
+void DumpH5MDParallel::writeId(hid_t fileId)
+{
+    using Datatype = int64_t;
+    constexpr int64_t dimensions = 1;  ///< dimensions of the property
+
+    std::string groupName = "/particles/atoms/" + idDataset;
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:133]: Let's keep string `/particles/atoms/` as a variable - for safety

File: src/io/DumpH5MDParallel.cpp [hunk 22]
```diff
@@ -0,0 +1,479 @@
...
+    for (iterator::CellListIterator cit(system_->storage->getRealCells()); !cit.isDone(); ++cit)
+    {
+        data.emplace_back(cit->id());
+    }
+    CHECK_EQUAL(int64_c(data.size()), numLocalParticles * dimensions);
+
+    std::vector<hsize_t> localDims = {1, uint64_c(numLocalParticles), dimensions};
+    std::vector<hsize_t> globalDims = {1, uint64_c(numTotalParticles), dimensions};
+
+    std::string dataset_name = "/particles/atoms/" + idDataset + "/value";
```
Comment (jkrajniak) [inline src/io/DumpH5MDParallel.cpp:147]: You could use already defind `groupName` variable

File: doc/sphinx/espressopp.io.DumpH5MDParallel.rst [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+H5MD_ is a format specification on top of the HDF5_ file format.
+
+.. _H5MD: https://nongnu.org/h5md/
+.. _HDF5: https://www.hdfgroup.org/solutions/hdf5/
+
+Dump Simulation
+---------------
+
+Dumping a simulation with ``espressopp.io.DumpH5MDParallel``
```
Comment (jkrajniak) [inline doc/sphinx/espressopp.io.DumpH5MDParallel.rst:12]: Am I right that this is only to dump a single frame during the simulation?

File: src/io/DumpH5MDParallel.py (hunk 17 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.py:34]: Will fix it. At some point we will have a PEP8 formatter :)

File: src/io/DumpH5MDParallel.cpp (hunk 22 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp:147]: fixed

File: src/io/DumpH5MDParallel.cpp (hunk 21 above)
Comment (XzzX) [reply src/io/DumpH5MDParallel.cpp]: introduced a new variable for 'atoms'

File: doc/sphinx/espressopp.io.DumpH5MDParallel.rst (hunk 19 above)
Comment (XzzX) [reply doc/sphinx/espressopp.io.DumpH5MDParallel.rst:30]: done

File: src/CMakeLists.txt (hunk 16 above)
Comment (XzzX) [reply src/CMakeLists.txt]: I will ship a custom `FindHDF5` until we enforce cmake > 3.19

File: CMakeLists.txt (hunk 15 above)
Comment (XzzX) [reply CMakeLists.txt:3]: There was already an attempt to do this:
https://gitlab.kitware.com/cmake/cmake/-/merge_requests/4534
TL;DR Enable C language if you use C language bindings.

**Comment by XzzX** [comment]
Ok, since there are no more comments about this merge request we should merge it and revisit the HDF5 output when somebody is actually using it.

@junghans Can you have a final look at it?

**Review by junghans — APPROVED** [review]

File: doc/sphinx/espressopp.io.DumpH5MDParallel.rst (hunk 23 above)
Comment (jkrajniak) [reply doc/sphinx/espressopp.io.DumpH5MDParallel.rst:12]: Ok, so to be clear, the following code:

```python
for step in range(100):
    integrator.run(100)
    h5mdump.dump()
```
will result with 100 frames or only a single last frame? If there will be only a single frame then I would put this information explicitly in documentation. Other dumps like xyz, pdb or gro behave differently, storing multiple frames.
